tools/mkcomposefs.c

```diff
@@ -760,46 +762,27 @@ static int mkdir_parents(const char *pathname, int mode)
 	return 0;
 }
 
-static int write_to_fd(int fd, const char *content, ssize_t len)
+static int copy_file_data(int sfd, int dfd)
 {
-	ssize_t res;
-
-	while (len > 0) {
-		res = write(fd, content, len);
-		if (res < 0 && errno == EINTR)
-			continue;
-		if (res <= 0) {
-			if (res == 0) /* Unexpected short write, should not happen when writing to a file */
-				errno = ENOSPC;
-			return -1;
-		}
-		len -= res;
-		content += res;
-	}
+	struct stat stat;
 
-	return 0;
-}
+	if(fstat(sfd, &stat) == -1)
+		return -1;
 
-#define BUFSIZE 8192
-static int copy_file_data(int sfd, int dfd)
-{
-	char buffer[BUFSIZE];
-	ssize_t bytes_read;
-
-	while (true) {
-		bytes_read = read(sfd, buffer, BUFSIZE);
-		if (bytes_read == -1) {
-			if (errno == EINTR)
-				continue;
-			return -1;
-		}
+	off_t len, ret;
+	len = stat.st_size;
 
-		if (bytes_read == 0)
+	do 
+	{
+		ret = copy_file_range(sfd, NULL, dfd, NULL, len, 0);
```

Comment:
I like this change, but it should be a separate independent commit with commit messages describing it.

Reply:
Added a separate commit just for this change